content/articles/ansible-and-nix.md

```diff
@@ -0,0 +1,332 @@
+    ];
+
+    src = ./.;
+}
+```
+
+However, if your dependencies are not in the nixos channel, you will need to
+declare them manually.
+
+For example the requests python library is declared like this:
+
+```nix
+requests = buildPythonPackage rec {
+  name = "requests-1.2.3";
+
+  src = pkgs.fetchurl {
+    url = "http://pypi.python.org/packages/source/r/requests/${name}.tar.gz";
+    md5 = "adbd3f18445f7fe5e77f65c502e264fb";
+  };
+
+  meta = {
+    description = "An Apache2 licensed HTTP library, written in Python, for human beings";
+    homepage = http://docs.python-requests.org/en/latest/;
+  };
+};
+```
+
+Which means than when building the package you are still dependent on pypi
+being up, unless the package is already cached.
+You can notice that the checksum of the archive is checked, so Nix will
+only build if the source package hasn't changed.
+
+Compared to a `requirements.txt` file, this is clearly more work. Fortunately
+the pypi2nix can take a `requirements.txt` file and write the corresponding
+derivations.
+
+Ansible doesn't have a single way of the handling of application dependencies.
+A common technique for a Python application is to install system packages with
+the distribution packages and python dependencies through pip.
+
+In the case of the voting test project this looks like this:
+
+```yaml
+    - name: global packages
+      apt: name={{item}} state=installed
+      with_items:
+        - python-virtualenv
+        - git
+        - uwsgi
+        - uwsgi-plugin-python
+        - nginx
+      sudo: yes
+
+    - name: python dependencies
+      pip: name=django version=1.6.6
+      sudo: yes
+```
+
+Of course, nothing stops you from building distribution packages locally and
+send them to your servers with Ansible.
+
+### Provisioning
+
+The general principle around provisioning is to ask for a set of servers to be
+running on which to deploy the application. A configuration management
+provisioning tool should start server that are not already running and re-use
+the ones that are.
+
+To keep track of the resources created by a deployment configuration, NixOps
+stores their name, state and access keys in a `state` file.
+Then when deploying, NixOps first matches the resources instantiated in the
+cloud provider to the state file and applies the changes needed to be in the
+state declared by the configuration.
+
+<blockquote><h4>TIP</h4>Don't include the test server and the production server
+of a project in the same state file.</blockquote>
+
+NixOps doesn't provide a way to share this state file across machines so the
+currently best solution is to [deploy from a shared
+machine](http://lists.science.uu.nl/pipermail/nix-dev/2015-February/016322.html).
+This is the main difficutly we encountered with NixOps as a shared machine is a
+single point of failure in terms of availability and security. Also, this
+forces us to somehow first deploy the shared deploy machine and then the other
+servers from there.
```

Comment:
I did think that we could check the sqlite file into git but we'd have to be careful because it contains secrets.